horovod/_keras/__init__.py

```diff
@@ -60,22 +40,42 @@ def get_gradients(self, loss, params):
             In DistributedOptimizer, get_gradients() is overriden to also
             allreduce the gradients before returning them.
             """
+            self._fail_if_executing_eagerly()
             gradients = super(self.__class__, self).get_gradients(loss, params)
             if hvd.size() > 1:
-                return self._allreduce_grads(gradients)
+                averaged_gradients = []
+                with tf.name_scope(self._name + "_Allreduce"):
+                    for grad in gradients:
+                        if grad is not None:
+                            if self._sparse_as_dense and \
+                                    isinstance(grad, tf.IndexedSlices):
+                                grad = tf.convert_to_tensor(grad)
+                            avg_grad = hvd.allreduce(grad,
+                                                     device_dense=self._device_dense,
+                                                     device_sparse=self._device_sparse,
+                                                     compression=self._compression)
+                            averaged_gradients.append(avg_grad)
+                        else:
+                            averaged_gradients.append(None)
+                    return averaged_gradients
             else:
                 return gradients
 
+        def apply_gradients(self, *args, **kwargs):
+            self._fail_if_executing_eagerly()
+            return super(self.__class__, self).apply_gradients(*args, **kwargs)
+
+        def _fail_if_executing_eagerly(self):
+            if hvd._executing_eagerly():
```

Comment:
Does this behavior differ between TF 2.0 and older versions?

Reply:
In order versions you need to opt in to eager mode. @romerojosh has a good point that eager mode is slower than graph mode, so if somebody is throwing a lot of resources at a problem, they should switch to graph mode first.